File: python system/inventory_system/repositories/inventory_repository.py
from typing import List, Optional
from models.inventory_item import InventoryItem
from repositories.database import InMemoryDatabase


class InventoryRepository:
    def __init__(self, db: InMemoryDatabase):
        self._db = db

    def get_all(self) -> List[InventoryItem]:
        return self._db.items

    def find_by_id(self, id: str) -> Optional[InventoryItem]:
        return self._db.get_item_by_id(id)

    def find_by_sku(self, sku: str) -> Optional[InventoryItem]:
        return self._db.get_item_by_sku(sku)

    def get_low_stock(self, threshold: int) -> List[InventoryItem]:
        return [item for item in self._db.items if item.quantity <= threshold]

    def save(self, item: InventoryItem) -> None:
        self._db.upsert_item(item)

    def delete(self, id: str) -> bool:
        return self._db.delete_item(id)

    def exists(self, id: str) -> bool:
        return self._db.get_item_by_id(id) is not None

    def sku_exists(self, sku: str) -> bool:
        return self._db.get_item_by_sku(sku) is not None

File: dart/transaction_system/dart_application_1/bin/inventory_repository.dart
// lib/repositories/inventory_repository.dart

import 'models/inventory_item.dart';
import 'database.dart';

class InventoryRepository {
  final InMemoryDatabase _db;

  InventoryRepository(this._db);

  List<InventoryItem> getAll() => _db.items;

  InventoryItem? findById(String id) => _db.getItemById(id);

  InventoryItem? findBySku(String sku) => _db.getItemBySku(sku);

  /// Returns all items with quantity at or below [threshold].
  List<InventoryItem> getLowStock(int threshold) =>
      _db.items.where((i) => i.quantity <= threshold).toList();

  void save(InventoryItem item) => _db.upsertItem(item);

  bool delete(String id) => _db.deleteItem(id);

  bool exists(String id) => _db.getItemById(id) != null;

  bool skuExists(String sku) => _db.getItemBySku(sku) != null;
}
